imnodes.cpp

```diff
@@ -2044,111 +2051,126 @@ void EditorContextMoveToNode(const int node_id)
 
 void Initialize()
 {
-    g.canvas_origin_screen_space = ImVec2(0.0f, 0.0f);
-    g.canvas_rect_screen_space = ImRect(ImVec2(0.f, 0.f), ImVec2(0.f, 0.f));
-    g.current_scope = Scope_None;
+    g = ContextCreate();
+    g->canvas_origin_screen_space = ImVec2(0.0f, 0.0f);
+    g->canvas_rect_screen_space = ImRect(ImVec2(0.f, 0.f), ImVec2(0.f, 0.f));
+    g->current_scope = Scope_None;
 
-    g.current_pin_idx = INT_MAX;
-    g.current_node_idx = INT_MAX;
+    g->current_pin_idx = INT_MAX;
+    g->current_node_idx = INT_MAX;
 
-    g.default_editor_ctx = EditorContextCreate();
-    EditorContextSet(g.default_editor_ctx);
+    g->default_editor_ctx = EditorContextCreate();
+    EditorContextSet(g->default_editor_ctx);
 
     const ImGuiIO& io = ImGui::GetIO();
-    g.io.emulate_three_button_mouse.modifier = &io.KeyAlt;
+    g->io.emulate_three_button_mouse.modifier = &io.KeyAlt;
 
-    g.current_attribute_flags = AttributeFlags_None;
-    g.attribute_flag_stack.push_back(g.current_attribute_flags);
+    g->current_attribute_flags = AttributeFlags_None;
+    g->attribute_flag_stack.push_back(g->current_attribute_flags);
 
     StyleColorsDark();
 }
 
 void Shutdown()
 {
-    EditorContextFree(g.default_editor_ctx);
-    g.editor_ctx = NULL;
-    g.default_editor_ctx = NULL;
+    EditorContextFree(g->default_editor_ctx);
+    ContextFree(g);
+    g = NULL;
 }
 
 void SetImGuiContext(ImGuiContext* ctx) { ImGui::SetCurrentContext(ctx); }
 
-IO& GetIO() { return g.io; }
+Context* ContextCreate()
+{
+    void* mem = ImGui::MemAlloc(sizeof(Context));
+    new (mem) Context();
+    return (Context*)mem;
+}
+void ContextFree(Context* ctx)
+{
+    if (ctx)
+        ImGui::MemFree(ctx);
+}
```

Comment:
We can probably do the exact same thing as ImGui does here, again making `Shutdown` an internal function:

```cpp
void ImGui::DestroyContext(ImGuiContext* ctx)
{
    if (ctx == NULL)
        ctx = GImGui;
    Shutdown(ctx);
    if (GImGui == ctx)
        SetCurrentContext(NULL);
    IM_DELETE(ctx);
}
```